pkg/controller/extensions/extension/extension.go

```diff
@@ -0,0 +1,251 @@
+/*
+Copyright 2019 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package extension
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	apps "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbac "k8s.io/api/rbac/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	controllerHandler "sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	corev1alpha1 "github.com/crossplaneio/crossplane/pkg/apis/core/v1alpha1"
+	"github.com/crossplaneio/crossplane/pkg/apis/extensions/v1alpha1"
+	"github.com/crossplaneio/crossplane/pkg/logging"
+)
+
+const (
+	controllerName = "extension.extensions.crossplane.io"
+
+	reconcileTimeout      = 1 * time.Minute
+	requeueAfterOnSuccess = 10 * time.Second
+
+	reasonCreatingRBAC       = "failed to create RBAC"
+	reasonCreatingDeployment = "failed to create deployment"
+)
+
+var (
+	log              = logging.Logger.WithName(controllerName)
+	resultRequeue    = reconcile.Result{Requeue: true}
+	requeueOnSuccess = reconcile.Result{RequeueAfter: requeueAfterOnSuccess}
+)
+
+// Reconciler reconciles a Instance object
+type Reconciler struct {
+	kube client.Client
+	factory
+}
+
+// Add creates a new Extension Controller and adds it to the Manager with default RBAC.
+// The Manager will set fields on the Controller and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	r := &Reconciler{
+		kube:    mgr.GetClient(),
+		factory: &extensionHandlerFactory{},
+	}
+
+	// Create a new controller
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to Extension
+	err = c.Watch(&source.Kind{Type: &v1alpha1.Extension{}}, &controllerHandler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
```

Comment:
Question: since this controller creates `Deployment` objects, wouldn't we want to requeue this extension object instance if the "child" deployment object was modified?
A good example in https://book.kubebuilder.io/basics/simple_controller.html

```suggestion
// Watch for changes to Deployments created by an Extension and trigger a Reconcile for the owner
  return errors.Wrapf(c.Watch(
		&source.Kind{Type: &apps.Deployment{}},
		&controllerHandler.EnqueueRequestForOwner{
			IsController: true,
			OwnerType:    &v1alpha1.Extension{},
		}), "failed to watch Deployment")
```

Reply:
i'll capture this watching on child artifacts to follow up on it.